Adds farneback optical flow algorithm

arijitkar98 opened the pull request:
This is the basic implementation of the algo. I will keep working on the correlation and pyramidal part.

arijitkar98 committed:
Adds farneback algo

arijitkar98 committed:
Modifies REQUIRE

arijitkar98 committed:
Merge branch 'master' into fb

arijitkar98 committed:
Adds tests for farneback optical flow

zygmuntszpak commented on src/farneback.jl:
```diff
@@ -0,0 +1,193 @@
+    a = ImageFiltering.Kernel.gaussian((σ,σ), (neighbourhood,neighbourhood))[-len:len, -len:len]
+
+    c = ones(size(img))
+    padded_c = reshape(padarray(c, Fill(0, (len, len)))[:], size(img) .+ 2*len)
```
If you want to work with a canonically indexed array instead of an `OffsetArray` then do this instead:
`padded_c = parent(padarray(c, Fill(0, (len, len))))`
`padded_img = parent(padarray(img, Fill(0, (len, len))))`

Documentation for `parent`. 
```julia
  parent(A)

  Returns the "parent array" of an array view type (e.g., SubArray), or the array itself if it is
  not a view.
```


Your current approach of vectorization and reshaping  allocates unnecessary memory.

zygmuntszpak commented:
Benchmarking this on one of test-cases shows that the code is allocating way too much memory inside critical loops resulting in really slow code:
```julia
 54.493782 seconds (145.13 M allocations: 95.058 GiB, 25.05% gc time)
``` 

I'm  experimenting with some changes to improve the performance.

zygmuntszpak commented:
If we assume that the confidence parameter is constant then we can move several key matrix multiplications outside of the main loop. After making such a change and removing some other unnecessary allocations, I've been able to substantially improve the benchmark results.:
```julia
  5.416388 seconds (34.69 M allocations: 3.721 GiB, 15.70% gc time)
```
This is still relatively slow. Our future work should probably focus on implementing the pyramidal convolution-based method in order to obtain good run-time performance. Our current implementation can at least serve as a reference of 'correctness' for when we attempt the convolution-based implementation. 

I'll post some of the requisite changes on our Slack channel so that you can incorporate them and modify this pull-request.

zygmuntszpak commented:
I believe I understand how to implement the convolution-based polynomial approximation method. They key equations are 4.9 and 4.10 on the page 27 (page 35 in the PDF) of the thesis [Multiscale curvature estimation polynomial expansion](http://liu.diva-portal.org/smash/get/diva2:312510/FULLTEXT01.pdf). 

I will work on it after you finalize the current pull-request.

arijitkar98 committed:
Modifies farneback to make it faster

(1) MVector --> SVector
(2) pinv() and svd() for 2x2
(3) Assumes c = 1 throughout
